### PR TITLE
Fixes stack context for json lines format that recovers from invalid JSON lines 

### PR DESCRIPTION
**Pull request by elstehle**
## Description

Addresses https://github.com/rapidsai/cudf/issues/14282.

For the JSON lines format that recovers after an invalid JSON line, we've had two issues when we were generating the stack context that is used downstream in the full JSON pushdown transducer.

For that format, we need to make sure that we "reset" the stack context after each JSON line. That is, 

1. We need to reset the stack to the empty stack after each JSON line, as the stack may not be empty after an erroneous  JSON line. E.g. `{"this opening brace is never closed":123\n{"<=this brace should be on the empty stack":...}`
2. We need to reset that we are outside of a string: `{"no matching end-quote on this line\n{"<=this quote is the beginning of a field name, not the end of the previous line's field name"`

This fixes above requirements as follows:
1. Was already implemented - but with an inappropriate scan operator that is not associative:
```
StackLevelT new_level = (symbol_to_stack_op_type(rhs.value) == stack_op_type::RESET)
                              ? 0
                              : (lhs.stack_level + rhs.stack_level);
```
E.g. (`{,n,{`,`},n,{`,`{,n,}`,`},n,}` all fail the associativity test). This was replaced with a `ScanByKey` that would start with a "fresh" stack level with each new key segment. 

2. Was addressed by changing the transition table of the finite-state transducer that filters out brackets and braces that are enclosed in quotes to go back to the `OOS` (`outside-of-string`) state after every newline. This behaviour requires that _every_ newline character is treated as a delimiter of a JSON line. This was confirmed by Spark Rapids, who is the targeted user for the recovery option to be the case.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by vuule — APPROVED** [review]
looks good. just one question to confirm my understanding

File: cpp/src/io/fst/logical_stack.cuh [hunk 1]
```diff
@@ -401,14 +476,41 @@ void sparse_stack_op_to_top_of_stack(StackSymbolItT d_symbols,
...
+      d_symbols,
+      detail::NewlineToResetStackSegmentOp<StackSymbolToStackOpTypeT>{symbol_to_stack_op});
+
+    rmm::device_uvector<StackSegmentT> key_segments{num_symbols_in, stream};
+    CUDF_CUDA_TRY(cub::DeviceScan::InclusiveSum(
+      temp_storage.data(),
+      total_temp_storage_bytes,
+      reset_segments_it,
+      thrust::make_transform_output_iterator(key_segments.data(),
+                                             detail::ModToTargetTypeOpT<StackSegmentT>{}),
```
[the remark this answers is not in the document]
Comment (elstehle) [reply cpp/src/io/fst/logical_stack.cuh:491]: Yup, right. We just need to make sure that two consecutive segments are assigned to two distinct key ranges, like:
```
...{"a":"b"}\n{"a":"b"}\n{"a"...
...999999999 0000000000 11111...
```
As an example where we wrap around after `9->0`

File: cpp/src/io/fst/logical_stack.cuh [hunk 2]
```diff
@@ -130,6 +139,36 @@ struct StackSymbolToStackOp {
...
+/**
+ * @brief Function object that wraps around for values that exceed the largest value of `TargetT`
+ */
+template <typename TargetT>
+struct ModToTargetTypeOpT {
+  template <typename T>
+  constexpr CUDF_HOST_DEVICE uint32_t operator()(T const& val) const
+  {
+    return val % (static_cast<T>(cuda::std::numeric_limits<TargetT>::max()) + static_cast<T>(1));
+  }
```
[the remark this answers is not in the document]
Comment (elstehle) [reply cpp/src/io/fst/logical_stack.cuh]: Thanks, good catch 👍 Fixed.

**Review by karthikeyann — APPROVED** [review]
Looks good to me.

**Pull request merged**